notebooks/Airline Delays - Final Writeup

```diff
@@ -251,6 +351,45 @@ ORDER BY origin_aa1_rain_depth"""))
+# MAGIC 
+# MAGIC The scope of this project was limited to model building. In future work, however, we could implement lab-based experiments to determine if our results are robust to greater scrutiny. And finally, we could deploy our model into a production environment with fault-tolerant infrastructure capable of ingestion for real-time predictions.
+# MAGIC 
+# MAGIC ### Bias-Variance Tradeoff
+# MAGIC 
+# MAGIC The bias-variance trade off describes the conflict between the two major sources of error in supervised machine learning that prevents algorithms from generalizing to unseen data. The error resulting from bias refers to error resulting from inadequete accounting of relevant relations between the input and output variables, also called underfitting, while the variance refers to error from over accounting of relations between input and output, also called overfitting.
+# MAGIC    
+# MAGIC Since we were working heavily with our training data before a final evaluation on the test set, we took steps to mitigate the risk of overfitting. Firstly, we used k-fold cross validation to help detect signs that we were overfitting as we optimized our model. Secondly, we adjusted the hyperparameters of our random forest model to avoid overfitting. By keeping a minimum tree depth, we limited the complexity of our models and discouraged it from modeling random noise in the training data. And by increasing the number of trees in the random forest, we increased the population of decision trees with low correlation to each other.
+# MAGIC 
+# MAGIC ### Spark and Parellel Frameworks
+# MAGIC   
+# MAGIC We were able to accomplish training and testing over tens of millions of records thanks to Spark and its parallel computing framework. Spark borrows many principles behind the functional programming paradigm to enable efficiency on large data sets. Below are some of the big data methodologies that allowed us to accomplish this project at scale.
+# MAGIC   
+# MAGIC ##### Higher order functions
+# MAGIC Where possible we leveraged Spark's built in functions such as filter and aggregate. When build in functions were not available to serve our needs, we used udf's applied as lambda functions in parallel on the data.
+# MAGIC   
+# MAGIC ##### High-level of Abstraction 
+# MAGIC Working at the Spark Dataframe level was the correct level of absraction for us in this project. It allowed for a quick workflow and pipeline build without concerning ourselves with low-level optimizations. In future work, however, it would likely be necessary to make lower level optimizations when our model was moved into deployment and production.
```

Comment:
doesn't have to be done before PR goes through but more of a note that we should add something about using SQL here. As Ramki said in class it offers an even higher level of abstraction from using dataframe functions directly and allows the framework to infer additional optimizations.